csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImpl.qll

```diff
@@ -198,21 +205,63 @@ module Private {
     boolean preservesValue
   ) {
     c.propagatesFlow(input, output, preservesValue)
+    or
+    // observe side effects of callbacks on input arguments
+    c.propagatesFlow(output, input, preservesValue) and
+    preservesValue = true and
+    isCallbackParameter(input) and
+    isContentOfArgument(output)
+    or
+    // flow from the receiver of a callback into the instance-parameter
+    exists(SummaryComponentStack s, SummaryComponentStack callbackRef |
```

Comment:
I considered this to be a general OO thing. Consider a call `x.apply(y)`, then the `x` plays two roles: it is used to determine the runtime dispatch target and it is being passed into the call as the instance argument. And since most languages have the concept of `this` being a parameter of member functions, then this feature felt pretty universal.
This could even be relevant for languages with lambdas that have no explicit way to reference themselves through a `this` parameter, since in most such cases lambdas support closures, which means that captured variables are actually stored in a closure object at the lambda construction point, which will likely need to be modelled as implicit instance fields on the implicit `this` of the lambda in order to get accurate flow.

Reply:
So actually not just an OO thing. This is relevant for purely functional programming as well, since it's needed for proper support of closures.

Reply:
OK, let's leave it in then.